mlir/lib/Quantum/Transforms/ScatterPatterns.cpp

```diff
@@ -0,0 +1,340 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "scatter"
+
+#include <algorithm>
+#include <vector>
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Index/IR/IndexOps.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+
+#include "mhlo/IR/hlo_ops.h"
+
+using namespace mlir;
+
+namespace {
+
+struct ScatterOpRewritePattern : public mlir::OpRewritePattern<mhlo::ScatterOp> {
+    using mlir::OpRewritePattern<mhlo::ScatterOp>::OpRewritePattern;
+
+    mlir::LogicalResult matchAndRewrite(mhlo::ScatterOp op,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        // Compute operation hash in case they are more than one scatter and they have different
+        // update function
+        auto opHash = OperationEquivalence::computeHash(op);
+        // Create the function to replace the update block from scatter
+        std::string funcName = "__catalyst_update_scatter" + std::to_string(opHash);
+
+        Location loc = op.getLoc();
+        // Add checks for supported cases (assumptions: no update windows dim, unique indices and
+        // sorted indices)
+        if (!op.getUniqueIndices() || !op.getIndicesAreSorted()) {
+            return failure();
```

Comment:
Returning failure would probably lead to passing the ScatterOp down the pipeline. I imagine that users would see kind of 'unkown op' errors then. Would it be better to call `emitError` with a verbose message explaining the failure reason?

Reply:
Yes I will add that